torch_geometric/contrib/nn/conv/graphwavenet.py

```diff
@@ -0,0 +1,252 @@
+        self.num_nodes = num_nodes
+
+        self.dropout = dropout
+
+        self.e1 = nn.Parameter(torch.randn(num_nodes, adptive_embeddings),
+                               requires_grad=True)
+        self.e2 = nn.Parameter(torch.randn(adptive_embeddings, num_nodes),
+                               requires_grad=True)
+
+        self.input = nn.Conv2d(in_channels=in_channels,
+                               out_channels=residual_channels,
+                               kernel_size=(1, 1))
+
+        self.tcn_a = nn.ModuleList()
+        self.tcn_b = nn.ModuleList()
+        self.gcn = nn.ModuleList()
+        self.gcn_adp = nn.ModuleList()
+        self.bn = nn.ModuleList()
+        self.skip = nn.ModuleList()
+
+        self.out_timesteps = out_timesteps
+        self.out_channels = out_channels
+
+        for d in dilations:
+            self.tcn_a.append(
+                nn.Conv2d(in_channels=residual_channels,
+                          out_channels=dilation_channels, kernel_size=(1, 2),
+                          dilation=d))
+
+            self.tcn_b.append(
+                nn.Conv2d(in_channels=residual_channels,
+                          out_channels=dilation_channels, kernel_size=(1, 2),
+                          dilation=d))
+
+            # GCNConv is used for performing graph convolutions over the
+            # normal adjacency matrix
+            self.gcn.append(
+                GCNConv(in_channels=dilation_channels,
+                        out_channels=residual_channels))
+
+            # Since the adaptive matrix is a softmax output, it represents a
+            # dense graph
+            # For fast training and inference, we use a DenseGCNConv layer
+            self.gcn_adp.append(
+                DenseGCNConv(in_channels=dilation_channels,
+                             out_channels=residual_channels))
+
+            self.skip.append(
+                nn.Conv2d(in_channels=residual_channels,
+                          out_channels=skip_channels, kernel_size=(1, 1)))
+            self.bn.append(nn.BatchNorm2d(residual_channels))
+
+        self.end1 = nn.Conv2d(in_channels=skip_channels,
+                              out_channels=end_channels, kernel_size=(1, 1))
+        self.end2 = nn.Conv2d(in_channels=end_channels,
+                              out_channels=out_channels * out_timesteps,
+                              kernel_size=(1, 1))
+
+    def forward(self, x, edge_index, edge_weight=None):
+        r"""
+        The forward pass for the GraphWaveNet
+        """
+        # Shapes:
+        # x -> (*, t, num_nodes, in_channels)
+        # edge_index -> (2, messages)
+        # edge_weight -> (messages)
+
+        # Both batched and unbatched input is supported
+        # In the case of a single batch, the input is broadcast
+        is_batched = True
+        if len(x.size()) == 3:
+            is_batched = False
+            x = torch.unsqueeze(x, dim=0)
+
+        batch_size = x.size()[:-3]
```

Comment:
Hi @wsad1, the `batch_size` parameter is supposed to signify batch size of training input, however it is for a single graph (as the model is tightly coupled to function on a single graph, due the nature of the self adaptive matrix being of fixed size). So this signifies several training examples on the same graph. This is to support fast training. 

For example, say we have the traffic speed at all nodes of a road network at 4pm-5pm (as 12 time steps of 5 minute windows each) and we predict the speeds for the window 5pm-6pm. Similarly consider we have another example from 8pm-9pm and want to predict the speeds for 9pm-10pm. These 2 training examples can be supplied as a batch of size 2, where each element in the batch has input of size (12, num_nodes, input_feature_size).

This can be thought of the conventional idea of batches in PyTorch NNs. My understanding is since the graph is same across time steps and all examples corresponding to the same graph, we do not need to handle batches in the manner in which it is done in ASAPPooling.  

Will be happy to clarify/discuss further! Thank you!

Reply:
@wsad1 hi sorry for the hiatus from my end, did not have the bandwidth to look into this, however I hope to wrap this up now.

To summarize the above discussion: In the code, I have used batch_size as an extra dimension, however PyG usually prefers a flattened input along with a batch vector to indicate which input corresponds to which batch. 

On contemplating, my proposal is to get rid of batching here completely. This is because PyG batching is usually representative of multiple graphs supported in the same forward pass. However wrt GraphWaveNet, we usually only deal with a single graph, and the batching I attempted here is for multiple training examples. Hence, batching in the PyG sense isn't really required here. 

So my proposal is that I can completely remove the batch dimension and only expect input of shape (timestep, num_vertices, num_features). 

Do you think this is reasonable? Will he happy to discuss further.